src/cargo/core/shell.rs

```diff
@@ -425,82 +398,43 @@ impl ShellOut {
         style: &Style,
         justified: bool,
     ) -> CargoResult<()> {
-        match *self {
-            ShellOut::Stream {
-                ref mut buffered_stderr,
-                ..
-            } => {
-                let mut buffer = buffered_stderr.buffer();
-                buffer.reset()?;
-                buffer.set_color(&to_termcolor_spec(*style))?;
-                if justified {
-                    write!(buffer, "{:>12}", status)?;
-                } else {
-                    write!(buffer, "{}", status)?;
-                    buffer.set_color(termcolor::ColorSpec::new().set_bold(true))?;
-                    write!(buffer, ":")?;
-                }
-                buffer.reset()?;
-                match message {
-                    Some(message) => writeln!(buffer, " {}", message)?,
-                    None => write!(buffer, " ")?,
-                }
-                buffered_stderr.print(&buffer)?;
-            }
-            ShellOut::Write(ref mut w) => {
-                if justified {
-                    write!(w, "{:>12}", status)?;
-                } else {
-                    write!(w, "{}:", status)?;
-                }
-                match message {
-                    Some(message) => writeln!(w, " {}", message)?,
-                    None => write!(w, " ")?,
-                }
-            }
+        let style = style.render();
+        let bold = (anstyle::Style::new() | anstyle::Effects::BOLD).render();
+        let reset = anstyle::Reset.render();
+
+        let mut buffer = Vec::new();
+        if justified {
+            write!(&mut buffer, "{style}{status:>12}{reset}")?;
+        } else {
+            write!(&mut buffer, "{style}{status}{reset}{bold}:{reset}")?;
         }
+        match message {
+            Some(message) => writeln!(buffer, " {message}")?,
+            None => write!(buffer, " ")?,
+        }
+        self.stderr().write_all(&buffer)?;
```

Comment:
obviously I'm just some outsider, but I'm surprised by how this PR adds many small allocations. The previous implementation seems to carefully avoid making those allocations.

It seems to me that some "code pain" in cargo is worth it given the huge number of daily cargo invocations.

In this particular case using a Cursor around an array seems like a simple alternative.

```suggestion
        let mut buffer = std::io::Cursor::new([0u8; 1024]);
        if justified {
            write!(&mut buffer, "{style}{status:>12}{reset}")?;
        } else {
            write!(&mut buffer, "{style}{status}{reset}{bold}:{reset}")?;
        }
        match message {
            Some(message) => writeln!(buffer, " {message}")?,
            None => write!(buffer, " ")?,
        }
        self.stderr()
            .write_all(&buffer.get_ref()[..buffer.position() as usize])?;
```

I'm not entirely sure why writing into a separate buffer before actually writing to stderr is required here. If it turns out that separate buffer is not needed, then this could work? 

```suggestion
        let stderr = self.stderr();

        if justified {
            write!(stderr, "{style}{status:>12}{reset} ")?;
        } else {
            write!(stderr, "{style}{status}{reset}{bold}:{reset} ")?;
        }

        if let Some(message) = message {
            write!(stderr, "{message}")?;
        }
```

obviously this is not the main point of this PR, it just surprised me. This is just one example, the same pattern seems to occur in a bunch of places and the allocations can be removed in similar ways.

Reply:
The original buffering was added in #12727 to workaround the fact that we aren't locking the stream (unsure why [lock](https://docs.rs/termcolor/latest/termcolor/struct.StandardStream.html#method.lock) wasn't used instead).

The difference between #12727  and this is that we aren't reusing the buffer.  

I expect allocations aren't our biggest concern here
- In most cases, we are writing to the screen which has its own performance concerns
- A good number of calls to these functions will already be having to do a `format!`.  I've found the biggest difference is between 'no allocation" and "allocation" code paths.  The difference between "3 allocations" and "4 allocations" is small.

If allocations were to be an issue, our options are
- Cache the `Vec` inside of `ShellOutput`
- Extend `anstream` to allow locking without transferring ownership
  - `AutoStream::lock` transfers ownership, which doesn't work here, because (1) `AutoStream` is stateful and it adds a lot of complexity in both the implementation and the caller (the locked stream needs a `&mut` back to the original stream) (2) most of the time none of this will matter but when it doesn, I found enough cases want access to an owned stream

Reply:
Since the number of logged status messages is relatively small, I don't think there's a going to be any measurable performance impact from this allocation. Using a fixed-length stack buffer would avoid the allocation, but also uses an unnecessary amount of stack space in most cases, and limits message length.